> For alternate distros, the failure seems to be in test infra
> 
> ```
>   File "/usr/lib/python3.14/site-packages/tox/run.py", line 20, in run
>     result = main(sys.argv[1:] if args is None else args)
>   File "/usr/lib/python3.14/site-packages/tox/run.py", line 39, in main
>     state = setup_state(args)
>   File "/usr/lib/python3.14/site-packages/tox/run.py", line 53, in setup_state
>     options = get_options(*args)
>   File "/usr/lib/python3.14/site-packages/tox/config/cli/parse.py", line 38, in get_options
>     guess_verbosity, log_handler, source = _get_base(args)
>                                            ~~~~~~~~~^^^^^^
>   File "/usr/lib/python3.14/site-packages/tox/config/cli/parse.py", line 58, in _get_base
>     from tox.plugin.manager import MANAGER  # load the plugin system right after we set up report  # noqa: PLC0415
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/python3.14/site-packages/tox/plugin/manager.py", line 11, in <module>
>     from tox.session.cmd.run import parallel, sequential
>   File "/usr/lib/python3.14/site-packages/tox/session/cmd/run/parallel.py", line 14, in <module>
>     from .common import env_run_create_flags, execute
>   File "/usr/lib/python3.14/site-packages/tox/session/cmd/run/common.py", line 19, in <module>
>     from tox.session.cmd.run.single import ToxEnvRunResult, run_one
>   File "/usr/lib/python3.14/site-packages/tox/session/cmd/run/single.py", line 11, in <module>
>     from tox.tox_env.python.virtual_env.package.pyproject import ToxBackendFailed
>   File "/usr/lib/python3.14/site-packages/tox/tox_env/python/virtual_env/package/pyproject.py", line 28, in <module>
>     from tox.tox_env.python.package import (
>     ...<5 lines>...
>     )
>   File "/usr/lib/python3.14/site-packages/tox/tox_env/python/package.py", line 13, in <module>
>     from .api import NoInterpreter, Python
>   File "/usr/lib/python3.14/site-packages/tox/tox_env/python/api.py", line 14, in <module>
>     from virtualenv.discovery.py_spec import PythonSpec
>   File "/usr/lib/python3.14/site-packages/virtualenv/discovery/py_spec.py", line 5, in <module>
>     from python_discovery import PythonSpec
> ModuleNotFoundError: No module named 'python_discovery'
> ```
> 
> For the integration test, not sure why this `ssh-import-id` is failing
> 
> ```
> E           AssertionError: 
> E           Found unexpected errors: ('ssh_import_id', ProcessExecutionError("Unexpected error while running command.\nCommand: ['sudo', '--preserve-env=https_proxy', '-Hu', 'ubuntu', 'ssh-import-id', 'gh:blackboxsw']\nExit code: 1\nReason: -\nStdout: -\nStderr: -"))
> ```

We fixed ssh-import-id with 207d38fa067a1dd265ab076bbc2f32e3b8bbe899. and the alpine CI jobs with 9c7d85d82c16c2379b1f004623c9d43507f31d3e.

--- a/cloudinit/net/dhcp.py
+++ b/cloudinit/net/dhcp.py
@@ -29,6 +29,7 @@ NETWORKD_LEASES_DIR = "/run/systemd/netif/leases"
 DHCLIENT_FALLBACK_LEASE_DIR = "/var/lib/dhclient"
 # Match something.lease or something.leases
 DHCLIENT_FALLBACK_LEASE_REGEX = r".+\.leases?$"
+NMCLI = "nmcli"
 UDHCPC_SCRIPT = """#!/bin/sh
 log() {
     echo "udhcpc[$PPID]" "$interface: $2"
@@ -145,6 +146,80 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
         if data.get(keyname):
             return data[keyname]
     return None
+
+
+def run_nmcli(nmcli_opts: List[str]) -> str:
+    nmcli_path = subp.which(NMCLI)
+    if not nmcli_path:
+        raise NoDHCPLeaseMissingDhclientError()
+
+    command = [nmcli_path] + nmcli_opts
+    try:
+        out, _ = subp.subp(
+            command,
+        )
+    except subp.ProcessExecutionError as error:
+        LOG.debug(
+            "nmcli command exited with code: %s stderr: %r stdout: %r",
+            error.exit_code,
+            error.stderr,
+            error.stdout,
+        )
+        raise NoDHCPLeaseError from error
+    return out
+
+
+def network_manager_load_leases(device: str) -> Dict[str, str]:
+    """Return a dictionary of lease options obtained from NM cli"""
+
+    opts = ["--fields", "DHCP4", "device", "show", device]
+    nmcli_out = run_nmcli(opts)
+
+    content = []
+    for line in nmcli_out.splitlines():
+        line = line.partition(":")[2].strip()
+        content.append(line)
+
+    return dict(configobj.ConfigObj(content, list_values=False))
+
+
+def find_correct_device_nmcli() -> Optional[str]:
+    """Return the lease value for
+    the first connected device as returned by 'nmcli'"""
+
+    device_list_opts = ["--terse", "device"]
+    nmcli_out = run_nmcli(device_list_opts)
+
+    for line in nmcli_out.splitlines():
+        if line == "":
+            continue
+        try:
+            dev_name, _, state, _ = line.split(":", 3)
+        except ValueError:
+            LOG.warning(
+                "Unexpected nmcli format: expected 4 colon-delimited"
+                " values, found %s",
+                line,
+            )
+            continue
+
+        # skip devices that are not connected
+        if state != "connected":
+            continue
+        # skip loopback
+        if dev_name == "lo":
+            continue
+        return dev_name
+    return None
+
+
+def network_manager_get_option_from_leases(keyname: str) -> Optional[str]:
+    leases = None
+    dev = find_correct_device_nmcli()
+    if dev:
+        leases = network_manager_load_leases(dev)
+
+    return leases.get(keyname) if leases else None
 
 
 class DhcpClient(abc.ABC):

--- a/cloudinit/sources/DataSourceCloudStack.py
+++ b/cloudinit/sources/DataSourceCloudStack.py
@@ -90,6 +90,7 @@ class DataSourceCloudStack(sources.DataSource):
         """Try obtaining a "domain-name" DHCP lease parameter:
         - From systemd-networkd lease (case-insensitive)
         - From ISC dhclient
+        - From network manager dhcp client
         - From dhcpcd (ephemeral)
         - Return empty string if not found (non-fatal)
         """
@@ -113,7 +114,19 @@ class DataSourceCloudStack(sources.DataSource):
 
         LOG.debug(
             "Could not obtain FQDN from ISC dhclient leases. Falling back to "
-            "%s",
+            "Network Manager leases"
+        )
+        with suppress(
+            dhcp.NoDHCPLeaseMissingDhclientError, dhcp.NoDHCPLeaseError
+        ):
+            domain_name = dhcp.network_manager_get_option_from_leases(
+                "domain_name"
+            )
+            if domain_name:
+                return domain_name.strip()
+
+        LOG.debug(
+            "Could not obtain FQDN from NM leases. Falling back to %s",
             self.distro.dhcp_client.client_name,
         )
         try:

--- a/tests/unittests/net/test_dhcp.py
+++ b/tests/unittests/net/test_dhcp.py
@@ -20,8 +20,11 @@ from cloudinit.net.dhcp import (
     NoDHCPLeaseInterfaceError,
     NoDHCPLeaseMissingDhclientError,
     Udhcpc,
+    find_correct_device_nmcli,
     maybe_perform_dhcp_discovery,
+    network_manager_load_leases,
     networkd_load_leases,
+    run_nmcli,
 )
 from cloudinit.net.ephemeral import EphemeralDHCPv4
 from cloudinit.subp import SubpResult
@@ -1392,3 +1395,84 @@ class TestMaybePerformDhcpDiscovery:
         with pytest.raises(NoDHCPLeaseInterfaceError):
             distro = mock.Mock(fallback_interface=None)
             maybe_perform_dhcp_discovery(distro, None)
+
+
+class TestNMDhcpLeases:
+    def test_find_correct_connected_device_first_match(self):
+        with mock.patch(
+            "cloudinit.net.dhcp.run_nmcli",
+            return_value=dedent(
+                """
+                   ens150:ethernet:unavailable:
+                   p2p-dev-wlp:wifi-p2p:disconnected:
+                   ens160:ethernet:connected:Wired connection 2
+                   ens256:ethernet:connected:Wired connection 3
+                   lo:loopback:connected (externally):lo
+                """
+            ),
+        ):
+            ret = find_correct_device_nmcli()
+        assert ret == "ens160"
+
+    def test_network_manager_load_leases(self):
+        expected_return = {
+            "broadcast_address": "172.16.127.255",
+            "dhcp_client_identifier": "01:00:0c:29:bf:c5:56",
+            "dhcp_lease_time": "1800",
+            "dhcp_server_identifier": "172.16.127.254",
+        }
+        with mock.patch(
+            "cloudinit.net.dhcp.run_nmcli",
+            return_value=dedent(
+                """
+                DHCP4.OPTION[1]: broadcast_address = 172.16.127.255
+                DHCP4.OPTION[2]: dhcp_client_identifier = 01:00:0c:29:bf:c5:56
+                DHCP4.OPTION[3]: dhcp_lease_time = 1800
+                DHCP4.OPTION[4]: dhcp_server_identifier = 172.16.127.254
+                """
+            ),
+        ):
+            ret = network_manager_load_leases("ens10")
+            assert ret == expected_return
+
+    @mock.patch("cloudinit.net.dhcp.subp.which", return_value="/usr/bin/nmcli")
+    @mock.patch("cloudinit.net.dhcp.subp.subp")
+    def test_run_nmcli(self, m_subp, m_which):
+        expected_return = dedent(
+            """
+                DHCP4.OPTION[1]: broadcast_address = 172.16.127.255
+                DHCP4.OPTION[2]: dhcp_client_identifier = 01:00:0c:29:bf:c5:56
+                DHCP4.OPTION[3]: dhcp_lease_time = 1800
+                DHCP4.OPTION[4]: dhcp_server_identifier = 172.16.127.254
+                """
+        )
+        m_subp.return_value = (expected_return, "")
+        ret = run_nmcli(["show"])
+        assert ret == expected_return
+
+    @mock.patch("cloudinit.net.dhcp.subp.which", return_value=None)
+    def test_run_nmcli_missing_nmcli(self, m_which):
+        """verify that absence of nmcli binary can result in
+        raising NoDHCPLeaseMissingDhclientError"""
+
+        with pytest.raises(NoDHCPLeaseMissingDhclientError):
+            run_nmcli(["show"])
+
+    @mock.patch("cloudinit.net.dhcp.subp.which", return_value="/usr/bin/nmcli")
+    @mock.patch("cloudinit.net.dhcp.subp.subp")
+    def test_run_nmcli_subp_err(self, m_subp, m_which):
+        """verify that when subp.ProcessExecutionError is raised while
+        running nmcli, it results in run_nmcli raising NoDHCPLeaseError"""
+
+        m_subp.side_effect = subp.ProcessExecutionError(exit_code=-5)
+
+        with pytest.raises(NoDHCPLeaseError):
+            run_nmcli(["--fields", "all", "device", "show"])
+
+        m_subp.assert_has_calls(
+            [
+                mock.call(
+                    ["/usr/bin/nmcli", "--fields", "all", "device", "show"],
+                ),
+            ]
+        )

--- a/tests/unittests/sources/test_cloudstack.py
+++ b/tests/unittests/sources/test_cloudstack.py
@@ -108,6 +108,10 @@ class TestCloudStackHostname:
             DHCP_MOD_PATH + ".networkd_get_option_from_leases",
             get_networkd_domain,
         )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
 
         with patch(
             MOD_PATH + ".util.load_text_file",

--- a/tests/unittests/sources/test_cloudstack.py
+++ b/tests/unittests/sources/test_cloudstack.py
@@ -67,6 +67,7 @@ class TestCloudStackHostname:
         self.hostname = "vm-hostname"
         self.networkd_domainname = "networkd.local"
         self.isc_dhclient_domainname = "dhclient.local"
+        self.nm_domainname = "nm.local"
 
         get_hostname_parent = mock.MagicMock(
             return_value=DataSourceHostname(self.hostname, True)
@@ -145,6 +146,201 @@ class TestCloudStackHostname:
             result = cloudstack_ds._get_domainname()
         assert self.isc_dhclient_domainname == result
 
+    def test_get_domainname_network_manager(self, cloudstack_ds, mocker):
+        """
+        Test if DataSourceCloudStack._get_domainname()
+        gets domain name from nmcli lease information
+        """
+        nmcliop = f"DHCP4.OPTION[5]:        domain_name = {self.nm_domainname}"
+        get_networkd_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".networkd_get_option_from_leases",
+            get_networkd_domain,
+        )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
+
+        mocker.patch(
+            MOD_PATH + ".util.load_text_file",
+            return_value=None,
+        )
+
+        mocker.patch(
+            DHCP_MOD_PATH + ".find_correct_device_nmcli",
+            return_value="ens120",
+        )
+
+        with patch(
+            DHCP_MOD_PATH + ".run_nmcli",
+            return_value=dedent(
+                """
+     DHCP4.OPTION[1]:          broadcast_address = 172.16.127.255
+     DHCP4.OPTION[2]:          dhcp_client_identifier = 01:00:0c:29:bf:c5:56
+     DHCP4.OPTION[3]:          dhcp_lease_time = 1800
+     DHCP4.OPTION[4]:          dhcp_server_identifier = 172.16.127.254
+     """
+                + nmcliop
+                + """
+     DHCP4.OPTION[6]:          domain_name_servers = 172.16.127.2
+     DHCP4.OPTION[7]:          expiry = 1775029195
+     DHCP4.OPTION[8]:          ip_address = 172.16.127.135
+     DHCP4.OPTION[9]:          next_server = 172.16.127.254
+     """
+            ),
+        ):
+            result = cloudstack_ds._get_domainname()
+        assert self.nm_domainname == result
+
+    def test_get_domainname_nm_multi_conn_nic(self, cloudstack_ds, mocker):
+        """
+        Test if DataSourceCloudStack._get_domainname()
+        can handle multi connected nic environments.
+        """
+        domain_name = f"{self.nm_domainname}"
+        get_networkd_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".networkd_get_option_from_leases",
+            get_networkd_domain,
+        )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
+
+        mocker.patch(
+            MOD_PATH + ".util.load_text_file",
+            return_value=None,
+        )
+
+        mocker.patch(
+            DHCP_MOD_PATH + ".run_nmcli",
+            return_value=dedent(
+                """
+                   ens160:ethernet:connected:Wired connection 1
+                   ens256:ethernet:connected:Wired connection 2
+                   lo:loopback:connected (externally):lo
+                """
+            ),
+        )
+
+        with patch(
+            DHCP_MOD_PATH + ".network_manager_load_leases",
+            return_value={"domain_name": domain_name},
+        ):
+            result = cloudstack_ds._get_domainname()
+        assert self.nm_domainname == result
+
+    def test_get_domainname_nm_single_conn_nic(self, cloudstack_ds, mocker):
+        """
+        Test if DataSourceCloudStack._get_domainname()
+        can handle one connected nic environments.
+        """
+        domain_name = f"{self.nm_domainname}"
+        get_networkd_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".networkd_get_option_from_leases",
+            get_networkd_domain,
+        )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
+
+        mocker.patch(
+            MOD_PATH + ".util.load_text_file",
+            return_value=None,
+        )
+
+        mocker.patch(
+            DHCP_MOD_PATH + ".run_nmcli",
+            return_value=dedent(
+                """
+                   ens160:ethernet:connected:Wired connection 1
+                   lo:loopback:connected (externally):lo
+                   ens256:ethernet:unavailable:
+                """
+            ),
+        )
+
+        with patch(
+            DHCP_MOD_PATH + ".network_manager_load_leases",
+            return_value={"domain_name": domain_name},
+        ):
+            result = cloudstack_ds._get_domainname()
+        assert self.nm_domainname == result
+
+    def test_get_domainname_nm_no_conn_nic(
+        self, cloudstack_ds, mocker, caplog
+    ):
+        """
+        Test if DataSourceCloudStack._get_domainname()
+        can handle one connected nic environments.
+        """
+        get_networkd_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".networkd_get_option_from_leases",
+            get_networkd_domain,
+        )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
+
+        mocker.patch(
+            MOD_PATH + ".util.load_text_file",
+            return_value=None,
+        )
+
+        mocker.patch(
+            DHCP_MOD_PATH + ".run_nmcli",
+            return_value=dedent(
+                """
+                   lo:loopback:connected (externally):lo
+                   ens256:ethernet:unavailable:
+                """
+            ),
+        )
+
+        result = cloudstack_ds._get_domainname()
+        assert "Could not obtain FQDN from NM leases" in caplog.text
+        assert (
+            "No domain name found in any DHCP lease; returning empty"
+            in caplog.text
+        )
+        assert result == ""
+
+    def test_get_domainname_nm_nodhcp_lease_err(
+        self, cloudstack_ds, mocker, caplog
+    ):
+        """
+        Test if DataSourceCloudStack._get_domainname()
+        can handle NoDHCPLeaseError.
+        """
+        get_networkd_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".networkd_get_option_from_leases",
+            get_networkd_domain,
+        )
+        mocker.patch(
+            MOD_PATH + ".dhcp.IscDhclient.get_newest_lease_file_from_distro",
+            return_value=True,
+        )
+
+        mocker.patch(
+            MOD_PATH + ".util.load_text_file",
+            return_value=None,
+        )
+
+        mocker.patch(
+            DHCP_MOD_PATH + ".network_manager_get_option_from_leases",
+            side_effect=NoDHCPLeaseError,
+        )
+
+        cloudstack_ds._get_domainname()
+        assert "Could not obtain FQDN from NM leases" in caplog.text
+
     def test_get_hostname_non_fqdn(self, cloudstack_ds):
         """
         Test get_hostname() method implementation
@@ -188,6 +384,12 @@ class TestCloudStackHostname:
         mocker.patch(
             DHCP_MOD_PATH + ".networkd_get_option_from_leases",
             get_networkd_domain,
+        )
+
+        get_nm_domain = mock.MagicMock(return_value=None)
+        mocker.patch(
+            DHCP_MOD_PATH + ".network_manager_get_option_from_leases",
+            get_nm_domain,
         )
 
         mocker.patch(

